File: blog/code/2022-12-02-hungarian/hungarian.cpp
#include <string.h>
#include <algorithm>
#include <iostream>
#include <vector>
#include <cassert>
#include <climits>

constexpr int S = 0;
constexpr int T = 1;
constexpr int MAXN = 200;

using namespace std;

struct Edge {
  int i;
  int j;
  int w;
};

struct Node {
  int i;
  int side;
};

struct Queue {
  Queue() {
  }

  void clear() {
    begin_ = end_ = 0;
  }

  bool empty() {
    return begin_ == end_;
  }

  void put(Node node) {
    data_[end_] = std::move(node);
    inc(end_);
  }

  Node get() {
    assert (!empty());
    auto node = data_[begin_];
    inc(begin_);
    return node;
  }

  void inc(int &idx) {
    idx = (idx + 1) % MAXN;
  }

  int size() {
    return ((end_ + MAXN) - begin_) % MAXN;
  }

  int begin_ = 0;
  int end_ = 0;
  Node data_[MAXN];
};

class Hungarian {

  public:
    Hungarian() {
      for (int i = 0; i < MAXN; i++) {
        for (int j = 0; j < MAXN; j++) {
          adj_[S][i][j] = adj_[T][i][j] = 0;
        }
      }
    }

  void set_edges(std::vector<Edge> &edges, int n) {
    n_ = n;

    // Clear previous edges
    for (auto edge : edges_) {
      adj_[S][edge.i][edge.j] = adj_[T][edge.j][edge.i] = 0;
    }

    for (auto edge : edges) {
      adj_[S][edge.i][edge.j] = adj_[T][edge.j][edge.i] = edge.w;
    }
    edges_ = edges;
  }

  int solve() {
    for (int i = 0; i < n_; i++) {
      match_[S][i] = match_[T][i] = -1;
      v_[i] = 0;

      u_[i] = 0;
      for (int j = 0; j < n_; j++) {
        u_[i] = std::max(u_[i], adj_[S][i][j]);
      }
    }

    for (int t = 0; t < n_; t++) {
      expand_matching();
    }

    int s = 0;
    for (int i = 0; i < n_; i++) {
      s += u_[i] + v_[i];
    }
    return s;
  }

  void expand_matching() {
    candidates_.clear();
    for (int i = 0; i < n_; i++) {
      if (match_[S][i] == -1) {
        candidates_.put({.side = S, .i = i});
      }

      slack_[i] = INT_MAX;

      parent_[S][i] = parent_[T][i] = -1;

      visited_[S][i] = visited_[T][i] = false;

    }

    while (true) {
      bool augmented = visit_forest();
      if (augmented) {
        return;
      }
      change_duals();
    }
  }

  bool visit_forest() {
    while (!candidates_.empty()) {
      auto node = candidates_.get();
      auto side = node.side;
      auto c = node.i;

      if (visited_[side][c]) {
        continue;
      }

      visited_[side][c] = true;

      if (side == S) {
        visit_s(c);
      } else {
        bool augmented = visit_t(c);
        if (augmented) {
          return true;
        }
      }
    }

    return false;
  }

  void visit_s(int i) {
    for (int j = 0; j <n_; j++) {
      if (match_[S][i] == j) {
        continue;
      }

      int slack = u_[i] + v_[j] - adj_[S][i][j];
      if (slack < slack_[j]) {
        slack_[j] = slack;
        parent_[T][j] = i;

        // edge (i,j) is now in the forest. keep visiting
        if (slack == 0) {
          candidates_.put({.side = T, .i = j});
        }
      }
    }
  }

  bool visit_t(int j) {
    int i = match_[T][j];
    if (i == -1) {
      augment(j);
      return true;
    }

    parent_[S][i] = j;
    candidates_.put({.side = S, .i = i});
    return false;
  }

  void augment(int j) {
    while (j != -1) {
      int i = parent_[T][j];
      assert (i != -1);
      match_[T][j] = i;
      match_[S][i] = j;
      j = parent_[S][i];
    }
  }

  void change_duals() {
    bool added = false;
    int delta = INT_MAX;
    for (int i = 0; i < n_; i++) {
      if (slack_[i] > 0) {
        delta = std::min(delta, slack_[i]);
      }
    }

    for (int i = 0; i < n_; i++) {
      if (visited_[S][i]) {
        u_[i] -= delta;
      }

      if (visited_[T][i]) {
        v_[i] += delta;
      } else if (slack_[i] > 0) { // frontier edge
        slack_[i] -= delta;
        if (slack_[i] == 0) {
          added = true;
          candidates_.put({.side = T, .i = i});
        }
      }
    }
    assert (added);
  }

  private:
  int n_;
  std::vector<Edge> edges_;
  int adj_[2][MAXN][MAXN];
  int match_[2][MAXN];
  int parent_[2][MAXN];
  int u_[MAXN];
  int v_[MAXN];
  int slack_[MAXN];
  bool visited_[2][MAXN];
  Queue candidates_;
};

int main() {

  int to, from, t, cost;
  int n1, n2;

  scanf("%d",&t);

  Hungarian hungarian;

  while(t--){
    scanf("%d %d",&n1, &n2);

    int n = std::max(n1, n2);

    std::vector<Edge> edges;
    while(1){
      scanf("%d %d %d",&from,&to,&cost);
      if(from == 0 && to == 0) break;
      edges.push_back({
        .i = from - 1,
        .j = to - 1,
        .w = cost,
      });
    }
    hungarian.set_edges(edges, n);

    int r = hungarian.solve();
    printf("%d\n",r);
  }
  return 0;
}

File: blog/code/2022-12-02-hungarian/hungarian.py
from sys import stdin
from queue import Queue

# Sufficently large number
INF = 12341234
S, T = 0, 1

class Hungarian:
  def __init__(self, edges, n, m):
    self.n = max(n, m)
    self.build_adjacency_mat(edges)

  def build_adjacency_mat(self, edges):
    adj = [ [None]*self.n, [None]*self.n ]
    for i in range(self.n):
      adj[S][i] = [0]*self.n
      adj[T][i] = [0]*self.n

    for [s, t, w] in edges:
      adj[S][s][t] = adj[T][t][s] = w

    self.adj = adj

  def init_parent(self):
    self.parent = [ [None]*self.n, [None]*self.n ]

  def init_candidates(self):
    self.candidates = Queue()
    # every exposed node in S is a candidate
    for i in range(self.n):
        if self.match[S][i] is None:
          self.candidates.put((0, i))

  def init_visited(self):
    self.visited = [[False]*self.n, [False]*self.n]

  def init_match(self):
    self.match = [[None]*self.n, [None]*self.n]

  def init_u(self):
    u = [0]*self.n
    for i in range(self.n):
      for j in range(self.n):
        u[i] = max(u[i], self.adj[S][i][j])

    self.u = u

  def init_v(self):
    self.v = [0]*self.n

  def init_slack(self):
    self.slack = [INF]*self.n

  def expand_matching(self):
    self.init_candidates()
    self.init_parent()
    self.init_visited()
    self.init_slack()

    while True:
      augmented = self.visit_forest()
      if augmented:
        break

      self.change_duals()

  def visit_s(self, i):
    # update pi from edges not in match
    for j in range(self.n):
      if self.match[S][i] == j:
        continue

      slack = self.u[i] + self.v[j] - self.adj[S][i][j]
      if slack < self.slack[j]:
        self.slack[j] = slack
        self.parent[T][j] = i

        # edge (i,j) is now in the forest. keep visiting
        if slack == 0:
          self.candidates.put((1, j))

  def visit_t(self, j):
    i = self.match[T][j]
    # found an augmenting path
    if i is None:
      self.augment(j)
      return True

    self.parent[S][i] = j
    self.candidates.put((0, i))
    return False

  def visit_forest(self):
    while not self.candidates.empty():
      [side, c] = self.candidates.get()

      if self.visited[side][c]:
        continue

      self.visited[side][c] = True

      if side == S: # c in S
        self.visit_s(c)
      else: # c in T
        augmented = self.visit_t(c)
        if augmented:
          return True
    return False

  def augment(self, j):
    while j is not None:
      i = self.parent[T][j]
      self.match[T][j] = i
      self.match[S][i] = j
      j = self.parent[S][i]

  def change_duals(self):
    delta = min(v for v in self.slack if v > 0)

    for i in range(self.n):
      if self.visited[S][i]:
        self.u[i] -= delta

    for j in range(self.n):
      if self.visited[T][j]:
        self.v[j] += delta
      # frontier edge
      elif self.slack[j] > 0:
        self.slack[j] -= delta
        # note: self.parent[j] has been set
        # during the forest visit
        if self.slack[j] == 0:
          self.candidates.put((1, j))

  def solve(self):
    self.init_match()
    self.init_u()
    self.init_v()
    self.init_slack()

    # for each iteration we should increase the matching side
    for t in range(self.n):
      self.expand_matching()

    return sum(self.u) + sum(self.v)


Tc = int(stdin.readline())
for t in range(Tc):
    [n, m] = [int(x) for x in stdin.readline().split(' ')]

    edges = []
    while True:
      [u, v, w] = [int(x) for x in stdin.readline().split(' ')]
      if u == 0:
        break

      edges.append([u-1, v-1, w])

    h = Hungarian(edges, n, m)
    v = h.solve()

    print(v)

    # break
